### ISSUE TITLE
Trap with action-at-a-distance when using `does` on immutable objects

### ISSUE THREAD

**Issue by AlexDaniel** [perl6/doc#2128]
See https://github.com/rakudo/rakudo/issues/1959 for more info.

Here's a gist:
```perl6
my $x = ‘foo’ does 42;

say ‘foo’.Int # 42
```

**Comment by zoffixznet** [comment]
This has nothing to do with objects being immutable, but due to implementation-specific caching:

```
$ perl6 -e '("a" ~ "a") does 42; ("a" ~ "a").Int.say'
Cannot convert string to number: base-10 number must begin with valid digits or '.' in '⏏aa' (indicated by ⏏)
  in block <unit> at -e line 1
```

When you wrote literal `‘foo’` in both places it got cached and the same object was re-used, so when you mix stuff into it, it'll affect it in both places.

**Comment by AlexDaniel** [comment]
@zoffixznet yeah, but isn't the implementation (hypothetically) allowed to reuse other types of immutable objects? Or only Strs are so special?

**Comment by AlexDaniel** [comment]
> Or only Strs are so special?

Well no, same happens with Ints, for example.

**Comment by zoffixznet** [comment]
> but isn't the implementation (hypothetically) allowed to reuse other types of immutable objects

By that logic the behaviour for this statement would be undefined on language-level, as it relies on the implementation-dependent re-use of immutable objects:

```
$ perl6 -e 'say +Set.new: (42, "foo"), (42, "foo")'
2
```

> Or only Strs are so special?

I'd say specialty applies to value types (that's what objects that have same `.WHICH` if their contents are the same, correct?). However, I see that `Map` and `Set` are actually value types, yet I recall when `List` was made a value type, that concept was rejected:

```
$ perl6 -e 'class Foo {}; say Set.new: Map.new((foo => Foo.new)), Map.new((foo => Foo.new))'
set(Map.new((foo => Foo.new)) Map.new((foo => Foo.new)))

$ perl6 -e 'class Foo {}; say Set.new: Map.new((foo => 42)), Map.new((foo => 42))'
set(Map.new((foo => 42)))

$ perl6 -e 'class Foo {}; say Set.new: Set.new(42), Set.new(42)'
set(set(42))
```

-----

And even were specialty applied, I think it's imperative for `.clone` method to return a whole new object, even if cached version is available, which would allow for use of `but` even with such objects. That's currently isn't the case for `Set.new.clone`, which keeps returning the same object if there are no elements.

**Comment by AlexDaniel** [comment]
@zoffixznet I see, thanks. I attempted to fix the title, feel free to have another go at it. Note that this ticket is about `does`, and to me it looks like we should simply advise against it? To be honest I don't entirely feel like I understand the difference between does and but (not intuitively at least). The issue with empty set not being buttable is just a bug IMO, and it's somewhat irrelevant to this ticket.

**Comment by zoffixznet** [comment]
> difference between does and but 

Ignoring the magic stuff they *both* do where they convert `"foo"` into `role { method ::($obj.^name) { $obj }`, the difference is:

* `does` = `.^mixin`
* `but` = `.clone.^mixin`

**Comment by zoffixznet** [comment]
Summary of https://github.com/rakudo/rakudo/issues/1959 :

* We don't currently have a way to forbid does on immutable objects. We should design and implement one. So eventually `does` will likely blow up if used on immutable things, and so such use should be discouraged at the moment

Also:

> *jnthn:* It's fine to intern immutable literal objects. 
> *Zoffix:* What are literal objects?
> *jnthn:* Those that are dependent solely on information available at compile time, and immutable from then on.
> *Zoffix:* Would it be correct to say that "an implementation may intern objects that have the same .WHICH"?
> *jnthn:* That's seem fair, though in reality the implementation for interning things like Int and Str at present relies on knowing their semantics - in part because it needs to compile the setting, so we need some bootstrapping allowances.

(intern = cache it; basically)

------

So I think as a general, implementation-independent way to document things is we want to say that if two objects have the same `.WHICH`, the implementation may cache such an object and return the cached copy instead of creating a new one when user requests a creation and so mutating it with `does` has the possibility of mutating the exact same object, which would affect its use elsewhere in the program.

At the same time, `.clone` is meant to create a new object even for things that have same `.WHICH`, so using `but` to mixin into such things is safe. And we can show the example:

```
my $x = ‘10’ does 42;
say ‘10’.Int; # 42

my $y = ‘20’.clone does 42;
say ‘20’.Int; # 20
say $y.Int; # 42

my $z = ‘30’ but 42;
say ‘30’.Int; # 30
say $z.Int; # 42
```

And basically `but` saves you calling a `.clone`

**Comment by zoffixznet** [comment]
> It's fine to intern immutable literal objects. [...] (intern = cache it; basically)

P.S.: we might want to mention that fact in a few more places rather than just `does`. For example, it affects the `=:=` operator's results (compare `Set.new =:= Set.new` vs `set() =:= set()`; Rakudo currently interns `set()` with no elements, though IMO that's an implementation-specific detail that doesn't necessarily need to be documented in `set()` unless stated it's Rakudo-specific)

**Comment by JJ** [comment]
This is no longer true in 2019.11 for literals. I have edited the example to reflect what does happen.
